packages/compass-sidebar/src/modules/databases.ts

```diff
@@ -105,64 +118,78 @@ export default function reducer(
-      expandedDbList,
+      [action.connectionId]: {
+        ...state[action.connectionId],
+        filterRegex: action.filterRegex,
+        filteredDatabases: filterDatabases(
+          state[action.connectionId]?.databases ?? [],
+          action.filterRegex
+        ),
+        expandedDbList,
+      },
     };
-  }
-
-  if (action.type === CHANGE_DATABASES) {
+  } else if (action.type === CHANGE_DATABASES) {
     return {
       ...state,
-      databases: action.databases,
-      filteredDatabases: filterDatabases(action.databases, state.filterRegex),
+      [action.connectionId]: {
+        ...state[action.connectionId],
+        databases: action.databases,
+        filteredDatabases: filterDatabases(
+          action.databases,
+          state[action.connectionId]?.filterRegex
+        ),
+      },
     };
   }
+
   return state;
 }
 
-/**
- * The change databases action creator.
- *
- * @param {Array} databases
- *
- * @returns {Object} The action.
- */
-export const changeDatabases = (databases: Database[]) => ({
+export const changeDatabases = (
+  connectionId: ConnectionInfo['id'],
+  databases: Database[]
+) => ({
   type: CHANGE_DATABASES,
+  connectionId,
   databases,
 });
 
 export const toggleDatabaseExpanded =
   (
-    id: string,
+    connectionId: ConnectionInfo['id'],
+    databaseId: string,
     forceExpand: boolean
   ): SidebarThunkAction<void, DatabasesAction> =>
   (dispatch, getState, { globalAppRegistry }) => {
-    const { database } = toNS(id);
-    const { databases } = getState();
-    const expanded = forceExpand ?? !databases.expandedDbList[database];
+    const { database } = toNS(databaseId);
+    const { expandedDbList } = getState().databases[connectionId];
+    const expanded = forceExpand ?? !expandedDbList[database];
+
     if (expanded) {
       // Fetch collections list on expand if we haven't done it yet (this is
       // relevant only for the code path that has global overlay disabled)
       globalAppRegistry.emit('sidebar-expand-database', database);
     }
-    dispatch({ type: TOGGLE_DATABASE, id: database, expanded });
+    dispatch({ type: TOGGLE_DATABASE, connectionId, database, expanded });
   };
 
 export const changeFilterRegex =
   (filterRegex: RegExp | null): SidebarThunkAction<void, DatabasesAction> =>
-  (dispatch, _getState, { globalAppRegistry }) => {
+  (dispatch, getState, { globalAppRegistry }) => {
     if (filterRegex) {
       // When filtering, emit an event so that we can fetch all collections. This
       // is required as a workaround for the syncronous nature of the current
       // filtering feature
       globalAppRegistry.emit('sidebar-filter-navigation-list');
     }
-    dispatch({
-      type: CHANGE_FILTER_REGEX,
-      filterRegex: filterRegex,
-    });
+
+    for (const connectionId of Object.keys(getState().databases)) {
+      dispatch({
+        type: CHANGE_FILTER_REGEX,
+        connectionId,
+        filterRegex,
+      });
+    }
```

Comment:
feels like this could be a single action, but we can refactor that in https://jira.mongodb.org/browse/COMPASS-7772

Reply:
Sounds good. My initial approach was this one because if for some reason one of the databases is not ready, it shouldn't affect others (as they are different dispatches). But I think we can achieve also this by using one single action.